_posts/2020-04-18-regression-chains/regression-chains.Rmd

```diff
@@ -0,0 +1,179 @@
+* Combine the input with the prediction of $y_{1}$, $\hat{y_{1}}$ and use this to predict $y_{2}$ within the second learner (i.e., $y_{2} \sim x_{1} + x_{2} + \hat{y_{1}} + y_{3}$).
+* Combine the input with the prediction of $y_{2}$ and use this to predict $y_{3}$ within the final learner (i.e., $y_{3} \sim x_{1} + x_{2} + \hat{y_{1}} + \hat{y_{2}}$).
+
+To combine predictions of a learner with the previous input, we rely on `PipeOpLearnerCV` and `PipeOpNOP` arranged in parallel via `gunion` combined via `PipeOpFeatureUnion`, e.g., in the first step:
+
+```{r}
+step1 = po("copy", outnum = 2, id = "copy1") %>>%
+  gunion(list(
+    po("learner_cv", learner = lrn("regr.lm"), id = "y1_learner"),
+    po("nop", id = "nop1")
+  )) %>>%
+  po("featureunion", id = "union1")
+step1$plot()
+```
+
+Within the second step we then have to define $y_{2}$ as the new target. This can be done using `PipeOpUpdateTarget` (note that `PipeOpUpdateTarget` by default drops the original target, here $y_{1}$). Typically one would use `PipeOpUpdateTarget` to apply a transformation function to the original target (possibly also renaming it) which must not be inverted later. However, in the special case of leaving the `trafo` at its default `identity` and providing a `new_target_name` that already matches a column of the input data, `PipeOpUpdateTarget` simply sets this column as the new target:
+
+```{r}
+#FIXME:
+mlr_pipeops$add("update_target", mlr3pipelines:::PipeOpUpdateTarget)
+```
+
+```{r}
+step2 = po("update_target", id = "y2_target",
+    param_vals = list(new_target_name = "y2")
+  ) %>>%
+  po("copy", outnum = 2, id = "copy2") %>>%
+  gunion(list(
+    po("nop", id = "nop2"),
+    po("learner_cv", learner = lrn("regr.lm"), id = "y2_learner")
+  )) %>>%
+  po("featureunion", id = "union2")
+```
+
+In the final third step we define $y_{3}$ as the new target:
+
+```{r}
+step3 = po("update_target", id = "y3_target",
+    param_vals = list(new_target_name = "y3")
+  ) %>>%
+  po("learner", learner = lrn("regr.lm"), id = "y3_learner")
+```
+
+The complete pipeline, more precisely `Graph`, looks like the following:
+
+```{r}
+graph = step1 %>>% step2 %>>% step3
+graph$plot()
+```
+
+## Evaluating the Pipeline
+
+By wrapping our `Graph` in a `GraphLearner`, we can perform `3-fold cross-validation` and get an average of the root-mean-square error (of course, in a real world setting splitting the data in a training and test set should have been done):
+
+```{r}
+rr = resample(task, learner = GraphLearner$new(graph),
+  resampling = rsmp("cv", folds = 3))
+rr$aggregate(msr("regr.mse"))
+```
+
+We can compare this to a simple linear regression model only predicting $y_{3}$:
+
+```{r}
+task_y3 = TaskRegr$new("y3", backend = dat, target = "y3")
+rr_y3 = resample(task_y3, learner = lrn("regr.lm"),
+  resampling = rsmp("cv", folds = 3))
+rr_y3$aggregate(msr("regr.mse"))
+```
```

Comment:
Could you, just for completeness, showcase, how, and if this works on a completely new dataset that only contains `x1-x3`? 
i.e. `predict(glrn, newdata= data.table(...))` or something?